Add validation.

mandyvenables opened the pull request:
Tested locally, with the internal test suite, main branch of project and client branch of project.

mandyvenables committed:
Add validation.

mandyvenables committed:
Add `additionalProperties: false`.

mandyvenables committed:
Add properties.

mandyvenables committed:
Add Changelog.

mandyvenables commented on schemas/bedrock-authn-token-http.js:
```diff
@@ -0,0 +1,170 @@
+ */
+'use strict';
+
+const postTokenQuery = {
```
@dlongley I did not see this come through in any of the tests, what is the expected input here?

dlongley replied:
It looks like maybe that's old and we don't need it? If you can double check that `req.query` isn't used where this would be applied ... and any other place where `req.query` *is* used we have a query validator for every value that is pulled out of it, then this can be removed.

mandyvenables replied:
Removed.

dlongley commented on package.json:
```diff
@@ -23,6 +23,7 @@
   },
   "homepage": "https://github.com/digitalbazaar/bedrock-authn-token-http",
   "dependencies": {
+    "bedrock-validation": "^5.5.0",
```
Bedrock libs should always be peer dependencies. Adding them as regular dependencies can result in duplicate libs being installed and multiple libs being loaded into memory -- breaking applications. Bedrock apps expect there to be exactly 1 version of any bedrock lib as they are stateful plugins into a bedrock framework that communicates across shared modules (bedrock events, etc.) -- as opposed to just regular code modules that provide isolated functions.

```suggestion
```

mandyvenables replied:
Updated

dlongley commented on schemas/bedrock-authn-token-http.js:
```diff
@@ -0,0 +1,170 @@
+const postToken = {
+  title: 'postToken',
+  type: 'object',
+  additionalProperties: false,
```
Each of these schemas needs to have `required` / `anyOf` added as appropriate to ensure that validation fails if the required fields are missing.

dmitrizagidulin reviewed (approved):
Looks reasonable. (I don't know much about the post token query part, if it's still used etc.)

mandyvenables committed:
Add required.

dlongley commented on schemas/bedrock-authn-token-http.js:
```diff
@@ -0,0 +1,171 @@
+  anyOf: [
+    {required: ['email']},
+    {required: ['account']}
+  ],
```
Aren't there other properties that are required as well? Looking at the route handler at a quick glance it seems it expects `authenticationMethod`, for example. If you make all of these required, what happens? It seems maybe only `hash` should not be required? We should require clients to set as many of these params as possible to ensure that they are doing what they want; leaving out only those that would create some kind of conflict.

dlongley replied:
We also want to try our best to fail closed here because it's an authn / security API -- we don't want someone to accidentally leave out `requiredAuthenticationMethods` when they really wanted to set it, for example.

mandyvenables replied:
`authenticationMethod` and `requiredAuthenticationMethods` are not used in any of the tests when calling `post.login`, are all of those incorrect? It does not seem like it's required as it works fine without it.

dlongley replied:
The tests are incorrect (is my guess).

mandyvenables replied:
Addressing the tests now, I pushed the update to test on the main branch of another repo.

dlongley commented on schemas/bedrock-authn-token-http.js:
```diff
@@ -0,0 +1,171 @@
+  anyOf: [
+    {required: ['email']},
+    {required: ['account']}
+  ],
```
Seems like `type` would be required here?

dlongley commented on schemas/bedrock-authn-token-http.js:
```diff
@@ -0,0 +1,171 @@
+    hash: {
+      type: 'string'
+    },
+    password: {
```
Do we really allow password here? Or only `hash`?

mandyvenables replied:
This test uses `password`, https://github.com/digitalbazaar/bedrock-authn-token-http/blob/main/test/mocha/10-api.js#L662
Should the test be updated?

dlongley commented on schemas/bedrock-authn-token-http.js:
```diff
@@ -0,0 +1,171 @@
+    account: {
+      type: 'string'
+    },
+    requiredAuthenticationMethods: {
```
We should require this.

dlongley commented on schemas/bedrock-authn-token-http.js:
```diff
@@ -0,0 +1,171 @@
+  required: ['type'],
+  anyOf: [
+    {required: ['email']},
+    {required: ['account']}
```
Aren't there some other combinations there? Like, if you use `type = totp` you also have to use `challenge` or `hash` or something like that? This doesn't seem strict enough.

mandyvenables replied:
I added in where some combination of both `account` or `email` and `hash` or `challenge` must be present.

dlongley commented on schemas/bedrock-authn-token-http.js:
```diff
@@ -0,0 +1,171 @@
+  title: 'getTokensQuery',
+  type: 'object',
+  additionalProperties: false,
+  required: ['email'],
```
Is type required?

mandyvenables replied:
Added

mandyvenables committed:
Update test and schemas.

dlongley reviewed (approved):
Approving to move things along -- only need that one last comment addressed (which requires a few tests to be updated).

mandyvenables committed:
Make 'authenticationMethod', 'requiredAuthenticationMethods' required.

mandyvenables committed:
Update tests.

dlongley committed:
Remove node 10.x from CI.

dlongley committed:
Remove erroneous query params.

dlongley committed:
Remove harded coded enums for authn methods.

- Authn methods are app-specific. Only one reserved rule exists
  that is understood by this module: `token-client-registration`.
- Remove requirement that required authn methods and authn method
  be set. This should be done but during a later breaking change.

dlongley committed:
Remove erroneous `token-client-registration` for user+pass auth.

- `token-client-registration` is an authn method for registering
  a client ID cookie, not password-based auth. Also, an authn method
  being used for a token should not match one that is required before
  the token can be authenticated (circular logic). Additionally, the
  password test here does not require that cookie to exist prior to
  the password being checked.

dlongley committed:
Remove erroneous `login-email-challenge` from test.

- The test is for password-based authn, not email code challenge.
  Additionally, the params being set were indicating that the
  password check was (erroneously) an email code check *and* that
  an email code check was required to be used prior to allowing
  the password to be checked. This is not possible (circular logic).
